### ISSUE TITLE
Ability to Persist Messages in External Stores

### ISSUE THREAD

**Issue by izzyacademy** [pydantic/pydantic-ai#530]
I am wondering if there is a plan to add ability to persist messages in remote stores like Redis or document stores instead of using plain memory to cache the messages [1]. There are scenarios where the application memory alone is not sufficient and we need to cache the state in a remote store and pick thing up later.

If this is already possible and I just have to add a plugin or dependency to do this, please let me know as well.

Thanks.

[1] https://ai.pydantic.dev/message-history/

**Comment by sydney-runkle** [comment]
I think this is something that could be supported in user code for the most part, though we'd be happy to consider support for loading externally stored message history.

**Comment by izzyacademy** [comment]
Thank you Sydney for sharing your thoughts. I appreciate it. 

I think it would be good to have a sort of interface or abstract parent class that users can implement or inherit from and extend to have their own persistent store (Redis, MongoDB, Postgres, MySQL etc) to cache these messages. This can then be injected in and used internally to cache the message history.

This will help with checkpointing and caching in the event of errors or restarts and we dont lose all the messages in the history so far. I will also simplify things so that the user does not have to spend too much energy to figure out this persistence implementation.

Those were my thoughts behind this suggestion.

I am open to collaborating to bring this into the framework

**Comment by samuelcolvin** [comment]
Hi @izzyacademy I agree we need this.

I think we need an ABC, with implementations for:
* memory
* file
* sqlite
* postgres

and an example of a custom implementation. I would need very strong evidence that people want another specific database before we implemented it.

I think it's best if I hav a crack at a first implementation of this, since I'll know how I want it to work.

**Comment by izzyacademy** [comment]
@samuelcolvin thanks for the update. I will wait for your initial draft and then I will share my feedback. After you start with the memory and SQL Lite, I can work on Redis and PostgreSQL. 

I believe there will be a strong need for Redis as the synchronous and async libraries have 45M and 800K downloads per month respectively. I will check back later on this.

**Comment by sandeep540** [comment]
Hi @izzyacademy  @samuelcolvin Even we are looking for similar type of implementation, Redis is preferred! for short-term memory, user-pref, user-details,  etc.

Crew AI has similar - https://docs.crewai.com/concepts/memory (short-term memory, long-term memory, entity memory, and contextual memory) 

I saw mem0, but it requires LLM, did not fit all our use-cases - https://github.com/mem0ai/mem0

**Comment by izzyacademy** [comment]
@sandeep540 thanks for your comments. I agree. Redis is a great key-value store for this use case.

**Comment by LeonidShamis** [comment]
I recently came across an [interesting video](https://www.youtube.com/watch?v=VKPngyO0iKg) by [Adam Lucek](https://www.youtube.com/@AdamLucek) about implementing different [memory types](https://www.psychologytoday.com/us/basics/memory/types-of-memory) that referenced two academic papers:

- [Cognitive Architectures for Language Agents, 2024](https://arxiv.org/pdf/2309.02427)
- [A Survey on the Memory Mechanism of Large Language Model based Agents](https://arxiv.org/pdf/2404.13501)

The types of memory that were implemented in that video are:

- Working Memory - Current conversation and immediate context
- Episodic Memory - Historical experiences and their takeaways
- Semantic Memory - Knowledge context and factual grounding
- Procedural Memory - The "rules" and "skills" for interaction

Could such memory types be considered for the API/ABC/implementation you are planning?

**Comment by izzyacademy** [comment]
Like the other external integrations, I think this could be [pydantic-ai-persistence] and we could integrate different data stores that implement the ABC.

Here are my initial thoughts based on the agent or graph workflow interactions with the persistent store

```python


from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any

@dataclass
class PersistentStore(ABC):
    """A persistent store ABC """

    conversation_id: str = None
    """This unique string is used to track different interactions with the persistent store. 
    
    Generate unique one is None is specfied
    """

    @abstractmethod
    async def append_entry(self, entry: Any):
        """Adds a record to the end of the list"""

    @abstractmethod
    async def append_entries(self, entries: list[Any]):
        """Adds records to the end of the list"""

    @abstractmethod
    async def prepend_entry(self, entry: Any):
        """Adds a record to the beginning of the list"""

    @abstractmethod
    async def get_all_entries(self) -> list[Any]:
        """Retrieves all the entries for this conversation"""

    @abstractmethod
    async def get_entries(self, start: int, end: int) -> list[Any]:
        """Retrieves a subset of the entries for this conversation"""

    @abstractmethod
    async def get_first_entry(self) -> Any:
        """Retrieves the first entry for this conversation"""

    @abstractmethod
    async def get_last_entry(self) -> Any:
        """Retrieves the last entry for this conversation"""

    @abstractmethod
    async def clear(self):
        """Wipes the list clean to start from an empty list"""

    @abstractmethod
    async def remove_first_entry(self) -> Any:
        """Removes the first entry for this conversation"""

    @abstractmethod
    async def remove_last_entry(self) -> Any:
        """Removes the last entry for this conversation"""

    @abstractmethod
    async def remove_entries(self, start: int, end: int) -> list[Any]:
        """Removes a subset of the entries for this conversation"""

    @abstractmethod
    async def count(self) -> int:
        """Returns the total num of messages for this conversation"""


class Memory(PersistentStore):
     pass

class PostgreSQLPersistence(PersistentStore):
    pass

class RedisPersistent(PersistentStore):
    pass




```

**Comment by slavakurilyak** [comment]
Mem0 uses two components to represent memory: (1) short-term memory, and (2) long-term memory. Traditional database for the former and vector database for the latter.

Google, on the other hand, uses (1) core, (2) contextual (long-term) memory, and (3) persistent memory, as outlined in their [Titans: Learning to Memorize at Test Time](https://arxiv.org/abs/2501.00663v1) paper. Why three you might ask?

> Inspired by human long-term memory system, **we design this memory module so an event that violates the expectations (being surprising) is more memorable**

**Comment by LeonidShamis** [comment]
> Here are my initial thoughts based on the agent or graph workflow interactions with the persistent store
> 
> from abc import ABC, abstractmethod
> from dataclasses import dataclass
> from typing import Any
> 
> @dataclass
> class PersistentStore(ABC):
>     """A persistent store ABC """
> 
>     conversation_id: str = None
>     """This unique string is used to track different interactions with the persistent store. 
>     
>     Generate unique one is None is specfied
>     """
> 
>     @abstractmethod
>     async def append_entry(self, entry: Any):
>         """Adds a record to the end of the list"""
> 
>     ...

I see similar `AgentMemory` usage with different `MemoryDB` implementations in another Agentic framework - [Agno](https://github.com/agno-agi/agno) (rebranded Phidata):

- [AgentMemory](https://github.com/agno-agi/agno/blob/6b3a96ed59bd312e9d58c5c93254a7b5af8180c2/libs/agno/agno/memory/agent.py#L54)

- [MemoryDB](https://github.com/agno-agi/agno/tree/main/libs/agno/agno/memory/db)

**Comment by Isopolito** [comment]
Hi, any updates on this work? I will be implementing memory strategies with our pydantic-ai agents soon and I don't want to reinvent the wheel if there's already a good solution available.

**Comment by Faylixe** [comment]
I am currently working on a Redis based state persistance plugin, I already finished the implementation and I just need to write and pass some tests to make it more reliable :)

https://github.com/opsie-dev/pydantic-graph-persistance-redis

I would gladly develop more if needed, also I would love to have some feedback on the initial implementation. Feel free to ping me when message history spec would be more fixed and i will add implementation for that as well :)

**Comment by aytekinar** [comment]
Hi @samuelcolvin,

Like @Faylixe, I have created a `pydantic-graph-persistence-postgresql` (as of now, private) repository that hosts a Python package to help get `pydantic-graph`s persisted on PostgreSQL.

I would like to understand where such integrations should live. Is the Pydantic (AI) team open to having a similar design/separation logic of `pydantic-graph` to that of `pydantic-ai` in that `pydantic-graph` can have, e.g., a "slim" version with only in-memory and file-based persistence implemented, whereas the external stores are supported through optional dependencies such as `redis` and/or `postgresql`? If yes, how can one start creating PRs and what should be the acceptence criteria to get the code into `pydantic-graph` (e.g., examples, documentation, testing)?

Looking forward to your guidance/suggestions...

Cc: @pinodeca and @AbeOmor

**Comment by dmontagu** [comment]
@aytekinar I think for us to upstream work on graph persistence, I'd want it to be compatible with the new beta API. The problem is that having parallel node execution makes it really hard to do graph persistence right.

Also, the most common/important use case for graph persistence is durable execution, which we now have support for through integrations with third-party tools like Temporal, DBOS, etc. Note that durable execution works well for both handling transient errors but also for patterns like human-in-the-loop.

The other obvious use case I'm aware of that isn't as straightforward to handle by just "throwing durable execution at the problem" is resuming a graph run from part-way through. (Though I will note that I believe at least temporal has machinery for this, though I am not very familiar with it; I suspect other durable execution providers have related functionality as well.) This is the strongest argument in favor of adding some form of more advanced graph/agent run persistence but I fear that trying to do that well would end up with us building a pydantic-ai-specific poor excuse for a durable execution framework, which I'm not keen to do.

In short, I think that to justify additional graph-run-persistence work, we would need to:
* Identify use cases that are not better-addressed through the use of existing durable execution frameworks
* Have a proposal for what it looks like to persist graph run state when the graph run involves parallel step execution

**Comment by sirianni** [comment]
Thanks @dmontagu for your thoughts.  I don't personally see a groundswelling of demand from the community for using a 'durable execution' framework here (but maybe I'm misreading it).  Perhaps this push is driven by a strategic partnership for Pydantic Logfire?  That's, of course, fine and healthy as OSS projects need to be funded.

[LangGraph checkpoints](https://docs.langchain.com/oss/python/langgraph/persistence) may be a relevant comparison to what's being asked for here.  Many users likely have existing queues/stores that they are comfortable with (e.g. Postgres, DynamoDB, Redis, etc.) and would prefer to leverage those vs. adopting new technology.

**Comment by dmontagu** [comment]
@sirianni It actually isn't related to Logfire, it's probably a wash but if anything I'd guess durable execution slightly _reduces_ the need for observability. Most of the people I have spoken to about this at conferences etc. have just want to use persistence or "checkpoint" like functionality as a way to deal with application failures — network requests failing, transient db issues, etc. — and durable execution frameworks are the "right" tool for that. Note that several of these tools are built specifically on top of postgres (e.g., [DBOS](https://github.com/dbos-inc/dbos-transact-py), https://github.com/earendil-works/absurd, etc.).

I am not opposed to offering some form of checkpointing functionality but it's really complicated to get it right when parallel execution is involved without massively sacrificing on ergonomics, and I just think durable execution is a fairly clear-cut solution for this that I expect is designed/maintained by full (third-party) teams to be more robust than something we tack onto the framework and only "battle test" against pydantic-ai use cases.

To be clear, I'm totally open to considering approaches to persistence/"snapshotting" graph execution state that do this successfully, I just think durable execution is the practical solution to most practical end-user challenges in the short (and possibly long) term.

---

The main challenge I see with making graph persistence work is — look at the implementation of `pydantic_graph.beta.graph._GraphIterator.iter_graph` and the state of that object (specifically, `state`, `active_tasks` and `active_reducers`) and think about how multiple tasks could be in flight that will modify the `state` object at different points in their execution. When and how do I snapshot things in a way that guarantees consistent execution behavior downstream of a given point in execution? We can snapshot the `active_reducers` and `active_tasks`, but really for consistent behavior, we need to snapshot precisely what is going on _inside_ the active_tasks — what lines of code have we reached and what updates to state have been performed, etc.? If a given task makes three changes to state, how many of them have happened by the time we reach a given point? Even if a given task makes only _one_ change to state, how do we ensure that change is executed precisely once when I am resuming from another snapshot? Like, when I took the last snapshot, had active task 1 already run far enough to make its modification to the state value, or not?

I thought about this hard when implementing — one option was to make it so that state can only be updated inside of handlers that are executed in serial (and tied to snapshotting). This has the downside of needing to add a bunch of fairly unintuitive structure to execution if you _aren't_ making use of snapshotting (specifically, having some new mechanism for specifying state updates), which either ends up losing all type safety and looking more or less like what langgraph does for state updates with reducer function annotations etc., or ends up introducing even more complex types and type-checking boilerplate than the library already has, which is already butting up against ergonomics imo (though, and admittedly I'm biased, I think the tradeoff is currently well worth it, but I couldn't come up with a version that was snapshotting-friendly that didn't make me want to puke).

And I'll note that while I'm not familiar with all the durable execution frameworks out there, ensuring you get consistent execution is **_precisely_** the use case of temporal — the sandbox ensures deterministic execution even of async tasks and the execution of the full graph run is deterministic other than the "activities", which are basically auto-checkpointed in the temporal database — it basically nukes the problem of execution consistency here.

**Comment by pgalilea** [comment]
Hi @izzyacademy @Isopolito, for my use cases, I ended with something like this: https://github.com/pgalilea/memx/blob/main/examples/py_ai_example.py
I tried to organize and package the utils in an extremely small library. The goal was to be extensible, framework-agnostic, and with sync/async support. I hope it helps!

**Comment by syumpx** [comment]
I built a thin adapter that solves this using [sayou](https://github.com/pixell-global/sayou) (a file-system inspired context store for AI agents) as the persistence backend:

**[sayou-pydantic-ai](https://github.com/pixell-global/sayou-pydantic-ai)** — gives any Pydantic AI agent persistent workspace storage + conversation history.

### Workspace tools (file storage, search, KV)

```python
from pydantic_ai import Agent
from sayou_pydantic_ai import SayouToolset

# Zero config — auto SQLite at ~/.sayou/, files persist across sessions
agent = Agent("openai:gpt-4o", toolsets=[SayouToolset()])
result = await agent.run("Save a note about our meeting decisions")
```

The agent gets 7 workspace tools (write, read, list, search, grep, glob, kv) through Pydantic AI's `AbstractToolset` interface. Every write creates a versioned file — nothing is lost.

### Conversation history persistence

```python
from sayou_pydantic_ai import SayouToolset, SayouMessageHistory
from sayou.workspace import Workspace

async with Workspace() as ws:
    history = SayouMessageHistory(ws, conversation_id="project-alpha")
    agent = Agent("openai:gpt-4o", toolsets=[SayouToolset(workspace=ws)])

    messages = await history.load()  # [] on first run
    result = await agent.run("Continue our discussion", message_history=messages)
    await history.save(result.all_messages())  # Persisted via KV store
```

Uses `TypeAdapter(list[ModelMessage])` for serialization. Supports `max_messages` truncation and `ttl_seconds` auto-expiry.

### Storage backends

- **Default**: SQLite (zero config, local development)
- **Production**: MySQL or any SQLAlchemy-compatible DB
- **Cloud**: S3 for file content + MySQL for metadata

```bash
pip install sayou-pydantic-ai
```

This doesn't replace a first-party persistence ABC (which I'd love to see), but it works today for anyone who needs persistence across agent sessions.

**Comment by AILIFE1** [comment]
For anyone needing this today: I've been using Cathedral (cathedral-ai.com) as the external store for pydantic-ai message history.

The pattern: serialize `model_messages` at session end into Cathedral memories, restore them at session start via `/wake`.

```python
import json
from cathedral import CathedralClient
from pydantic_ai import Agent

client = CathedralClient(api_key="...", agent_id="my-agent")

# Restore previous session
ctx = client.wake()
prior_messages = []
for m in ctx.memories:
    if m.category == "message_history":
        try:
            prior_messages = json.loads(m.content)
            break
        except:
            pass

agent = Agent("openai:gpt-4o")
result = await agent.run("continue from where we left off", message_history=prior_messages)

# Persist for next session
serialized = json.dumps([m.model_dump() for m in result.all_messages()])
client.remember(serialized, category="message_history", importance=0.9)
```

Cathedral stores the full serialized message history as a memory. On next session, `/wake` returns it, you deserialize and pass as `message_history`. Cross-session continuity without modifying pydantic-ai internals.

Free API, 1,000 memories/agent. Also `pip install cathedral-server` for a local SQLite-backed version with the same REST interface.

**Comment by projectmnemos** [comment]
## Summary

I'd like to propose an integration with **mnem** (https://github.com/Uranid/mnem), a Rust-native, content-addressed knowledge graph engine that functions as "git for agent memory" - giving agents a persistent, versioned, auditable memory layer with no external dependencies.

## The gap this fills

pydantic's current memory implementation is session-scoped, no versioning, requires cloud service, no audit trail. mnem addresses this with:

- **Versioned storage**: every fact write is a content-addressed commit (BLAKE3 + DAG-CBOR)
- **Three-way merge**: concurrent agent writes are reconciled without silent overwrite
- **Full audit trail**: `mnem blame`, `mnem diff`, `mnem log` on the knowledge graph
- **Hybrid retrieval**: HNSW vector + BM25/SPLADE sparse + multi-hop graph traversal, fused via RRF
- **Offline-first**: bundled ONNX embedder, single ~40MB binary, no API keys
- **MCP-native**: ships with a full MCP server - integrates with any MCP-capable host with zero
  framework-specific code

## Benchmark evidence

On published memory benchmarks, mnem scores:
- LoCoMo R@5 = 0.726 (+0.218 over MemPalace, 1986 questions)
- LongMemEval R@5 = 0.966 (500 questions)
- MemBench R@5 = 0.960
- ConvoMem recall = 0.976 (+0.047 over baseline)

## Proposed integration shape

The simplest integration path is through the existing MCP interface - mnem runs as a local MCP server and any MCP-capable agent host can read/write to it without framework-specific bindings. A tighter integration (e.g., a dedicated memory backend class) is also feasible if the team is interested.

## License

Apache-2.0. No cloud dependency, no managed service requirement.

## Quick start

```
cargo install --locked mnem-cli --features bundled-embedder

mnem serve  # starts HTTP + MCP server on [[localhost](http://localhost/)](http://localhost)

mnem integrate  # auto-wires into Claude Code, Cursor, Continue, Zed
```

Happy to discuss design options or submit a draft PR.

**Comment by visy-ani** [comment]
For anyone landing here looking for a managed option (instead of rolling Redis/Mongo persistence yourself), [`maximem-synap-pydantic-ai`](https://pypi.org/project/maximem-synap-pydantic-ai/) plugs [Synap](https://maximem.ai) (long-term memory layer) into pydantic-ai. Persistence is handled server-side; you just scope by `user_id` + `customer_id` and call `agent.run()` normally — message history persists and relevant context gets injected automatically.

```python
from pydantic_ai import Agent
from maximem_synap import MaximemSynapSDK
from synap_pydantic_ai import attach_synap_memory

sdk = MaximemSynapSDK(api_key="sk-...")
agent = Agent(...)
attach_synap_memory(agent, sdk=sdk, user_id="alice", customer_id="acme")
```

Open source: https://github.com/maximem-ai/maximem_synap_sdk/tree/main/packages/integrations/synap-pydantic-ai — related design discussion in #4773.

**Comment by IgnazioDS** [comment]
Sharing a PostgreSQL implementation I've been using in production with FastAPI + pydantic-ai, since I haven't seen a psycopg3-native example in this thread yet.

The core of what I built is a thin async class around a single `conversation_messages` table that stores serialized `ModelMessage` objects with `TypeAdapter`:

```python
from pydantic_ai.messages import ModelMessage, ModelMessagesTypeAdapter
from psycopg import AsyncConnection
from typing import Sequence
import orjson

class PostgresMessageStore:
    def __init__(self, conn: AsyncConnection, conversation_id: str):
        self._conn = conn
        self._conversation_id = conversation_id

    async def load(self) -> list[ModelMessage]:
        async with self._conn.cursor() as cur:
            await cur.execute(
                "SELECT message_data FROM conversation_messages "
                "WHERE conversation_id = %s ORDER BY created_at ASC",
                (self._conversation_id,)
            )
            rows = await cur.fetchall()
        if not rows:
            return []
        return ModelMessagesTypeAdapter.validate_json(
            orjson.dumps([orjson.loads(r[0]) for r in rows])
        )

    async def append(self, messages: Sequence[ModelMessage]) -> None:
        data = ModelMessagesTypeAdapter.dump_json(list(messages))
        parsed = orjson.loads(data)
        async with self._conn.cursor() as cur:
            await cur.executemany(
                "INSERT INTO conversation_messages (conversation_id, message_data) "
                "VALUES (%s, %s)",
                [(self._conversation_id, orjson.dumps(m)) for m in parsed]
            )
        await self._conn.commit()
```

Usage in an agent run:
```python
store = PostgresMessageStore(conn, thread_id)
history = await store.load()
result = await agent.run(user_input, message_history=history)
await store.append(result.new_messages())
```

A few things I ran into that are worth capturing in any first-party ABC design:

1. **`new_messages()` vs `all_messages()`**: Storing only `new_messages()` per turn and reconstructing the full history on load (from `ORDER BY created_at`) is more robust than storing the whole list on each turn. It avoids the double-storage problem and makes truncation easier.

2. **Serialization fidelity**: `ModelMessagesTypeAdapter` handles the discriminated union correctly for all message types including tool calls and tool results. Using plain `jsonable_encoder` or `model_dump()` loses the type discriminator and breaks deserialization. Any ABC implementation needs to use `ModelMessagesTypeAdapter` explicitly.

3. **Token budget trimming**: In production I have a separate `trim_to_token_budget(messages, max_tokens)` utility that truncates old messages before passing to the agent. This belongs outside the store (it's a retrieval concern, not persistence), but it's something the ABC design should leave room for — the store returns the full history, and callers decide how much to pass to the model.

Would be very happy to contribute a `PostgresMessageStore` implementation once the ABC shape is settled.

**Comment by ferhimedamine** [comment]
We ran into this exact problem at scale and the serialization part is actually the easy bit — the hard part is what happens when your persisted history grows to thousands of messages across sessions and you need to recall the *right* 20 for the current context window. Plain key-value persistence (Redis, Postgres JSONB) works until it does not, because you still need relevance-based retrieval on top. What ended up working for us is treating stored messages as memories with metadata (timestamps, access counts, importance scores) and running hybrid search — BM25 for keyword precision plus vector similarity for semantic recall — at query time. We open-sourced the Python SDK for this pattern: [dakera-py](https://github.com/Dakera-AI/dakera-py). The `store` / `recall` API maps almost 1:1 to the `save` / `load` shape people are building here, but with a ranked retrieval path so you are not blindly replaying entire histories.

**Comment by dsfaccini** [comment]
> This comment was posted by Claude Code using claude-opus-5 on behalf of David.

**A session is a `conversation_id` plus a durable ordered log of its messages.** We ship both endpoints and no wire: identity is resolved, inherited and forkable; `ModelMessagesTypeAdapter` is full fidelity; `capabilities` is the seam. No store type exists in `pydantic_ai_slim`, which is why this thread keeps producing hand-rolled stores that are all the same three methods.

**Closing pydantic-ai-harness#176 was right on substance, wrong on residue.** `StepPersistence` is more capable and was right for orchestration, but #176 also carried a user-facing facade (four methods, one opaque id, auto save and restore) that went with it. So the fix is additive: put the facade back over the better substrate.

**Recommended:** a three-method async `SessionStore` protocol plus an in-memory reference in core; every real backend stays in Harness behind a `StepStore` adapter. No new `Agent.run` argument, since `conversation_id` already means the dialogue and #4773's `session_id=` would name one concept twice.

**Seven forks. None decided.**

| # | Fork | Recommendation |
|---|---|---|
| 1 | Log or resumable run state | Log |
| 2 | Branching | Reserve a parent pointer; no branch methods |
| 3 | Stateless-server doctrine | Not contradicted; rewrite it in the same PR |
| 4 | dmontagu's objection | Live, cleared by a log, not the binding one |
| 5 | Sessions vs memory (#4773) | Both, layered; Harness ships all three tiers |
| 6 | Sync + async | Async only |
| 7 | Provider-managed state | Model setting, not a store backend |

**Findability ships separately** in #7821. It needs no decision and does not close this issue.

@DouweM on the concept. @adtyavrdhn on fork 5: your "not inclined to add a memory store to the core" on #4773 is the objection this must clear; the core/Harness split is the answer offered.

Unsurveyed: LangChain deep agents; `AdvancedSQLiteSession` branching. Detail below.

<details>
<summary><b>Field survey</b></summary>

**LangGraph.** Two deliberately separate abstractions: `BaseCheckpointSaver` for graph channel snapshots, `BaseStore` for cross-thread memory. Identity is `{thread_id, checkpoint_ns, checkpoint_id}`. Its SQLite primary key is `(thread_id, checkpoint_ns, checkpoint_id)`, and `parent_checkpoint_id` is a **nullable column, not part of the key**. Cautionary detail: `BaseCheckpointSaver` launched small and now carries roughly 12 sync methods plus a full async mirror, paid by every community backend forever.

**OpenAI Agents SDK.** `Session` is four async methods (`get_items`, `add_items`, `pop_item`, `clear_session`) published as both a `@runtime_checkable` Protocol and an ABC. Flat `session_id: str`. Eight backends ship. Three moves worth stealing: encryption and compaction are *decorator sessions* that wrap another session rather than flags on the protocol; branching lives in one backend (`AdvancedSQLiteSession`), not in the protocol; server-side state is modeled as just another backend.

**OpenAI platform.** Chaining does not save money: "Even when using `previous_response_id`, all previous input tokens for responses in the chain are billed as input tokens in the API." Responses are kept 30 days; conversation objects have no TTL. Unusable under Zero Data Retention.

**Anthropic Agent SDK.** Opaque and SDK-managed. The only surface is `resume=` / `fork_session=`.

**Vercel AI SDK.** Ships no store at all. Documents writing your own `loadChat` / `saveChat`, saving from `onEnd`.

**Agreed on:** one opaque id, one injection point, async-first, four to six narrow methods, opaque blob per item, long-term memory as a separate abstraction, branching and encryption kept out of the base protocol. **Diverges on:** the stored unit (graph state vs flat item list) and whether the framework ships backends at all.

</details>

<details>
<summary><b>What we ship today, and the precise gap</b></summary>

Core has `conversation_id` (resolved, inherited from history, forkable via `'new'`), `ModelMessagesTypeAdapter`, and the `capabilities` seam. It has no store. `pydantic_graph.persistence` was deleted in v2 (#5470), and `docs/migration.md` now routes those users to Harness, which is why this issue is live again.

Harness is more complete than it looks from here. `StepStore.list_runs` already filters by `conversation_id`; `StepEvent`, `ContinuableSnapshot` and `RunRecord` all carry it; four backends ship (in-memory, file, SQLite, MongoDB); and the multi-turn resume recipe is already documented. `ConversationSearch` already provides BM25 `retrieve(query, k)` over what `StepPersistence` stored. `Memory` already provides the namespaced notebook.

So the gap is not capability. It is address shape and findability. Resuming a conversation today costs: find Harness, find `StepPersistence` under Capabilities, learn `run_id` vs `conversation_id` vs `parent_run_id`, then write a two-line recipe over `list_runs` and `continue_run`. What people wanted was `load(conversation_id)`.

</details>

<details>
<summary><b>API sketch</b></summary>

```python
@runtime_checkable
class SessionStore(Protocol):
    async def get_messages(
        self, conversation_id: str, *, limit: int | None = None
    ) -> list[ModelMessage]: ...

    async def add_messages(
        self, conversation_id: str, messages: Sequence[ModelMessage]
    ) -> None: ...

    async def clear(self, conversation_id: str) -> None: ...
```

Three deliberate deviations from OpenAI's `Session`:

- **`conversation_id` is a parameter, not constructor state.** Binding one store object to one id forces a fresh object per web request; passing it per call lets one instance serve every conversation, which is what a connection pool wants.
- **No `pop_item`.** Popping is the one operation that makes the log mutable, and an append-only log is what a branchable tree later requires. Adding a method is not a break; removing one is.
- **Append, not save-all.** `add_messages(new_messages())` rather than `save(all_messages())`. This is the most-repeated lesson in this thread, stated best by @IgnazioDS from production: storing only `new_messages()` per turn "avoids the double-storage problem and makes truncation easier".

Usage, with no new `Agent.run` argument:

```python
agent = Agent(
    'openai:gpt-5.2',
    capabilities=[SessionPersistence(store=SqliteSessionStore('chat.db'))],
)

r1 = await agent.run('My name is Arjun.', conversation_id='user-123')
# new process, new request, history restored automatically
r2 = await agent.run("What's my name?", conversation_id='user-123')
```

One implementation constraint: #176's design used `before_run` to prepend restored messages. Core's `before_run` is now documented observe-only, so auto-restore has to go through `wrap_run` (what `StepPersistence` already does) or `before_model_request`.

</details>

<details>
<summary><b>Reasoning per fork</b></summary>

**1. Message log, not resumable run state.** dmontagu's objection is specifically to graph-state checkpointing under parallel node execution, and a message log stores no graph state. We already deleted the alternative in #5470. The field is four-to-one for item lists. These are not incompatible primary keys, as both address by conversation; they differ in stored unit, so a checkpoint could be added later as another entry kind. What is expensive to retrofit is replay semantics onto opaque blobs, so the entry record wants a kind discriminator from day one.

**2. Reserve the parent pointer, ship no branching.** Correcting a premise: `parent_checkpoint_id` is not in LangGraph's primary key, and OpenAI keeps branching out of `Session` entirely. So the field does not support putting branch methods in the protocol. What it does support is a parent pointer on the stored record from day one, because adding a nullable column to already-shipped third-party backends is a migration every implementer must perform. Cost: one nullable field. This also keeps pydantic-ai-harness#321's tree a pure superset rather than a divergence. Note we already branch at conversation level via `conversation_id='new'`.

**3. The stateless-server doctrine is not contradicted.** The paragraph in `docs/message-history.md` makes a narrower claim than "the server is stateless": it says a run is reconstructed from the history supplied with the request, so a client that can submit history can forge it. A session store does not contradict that, it is the remedy that paragraph already names ("persist paused runs server-side and resume them with your own `deferred_tool_results`"). Today we recommend that and ship nothing to do it with. The rewrite is an upgrade, and it belongs in the same PR as the store.

**4. dmontagu's position is live and cleared, but is not the binding objection.** His two bars were: use cases not better served by durable execution, and a story for parallel step execution. The second is cleared vacuously by a log. The first is cleared on the merits, and the two sides have been talking past each other: durable execution restarts a *failed run*; it does not hand a web app the *previous conversation* on a fresh request when no run is in flight. Reading all comments here, every commenter is doing load-history-at-request-time and not one asked for graph checkpointing. The more current objection is @adtyavrdhn's on #4773, which is why fork 5 is addressed to him.

**5. Sessions and memory are layered, not a choice.** Harness already ships all three tiers cleanly separated: `StepPersistence` (the durable record, keyed by conversation), `ConversationSearch` (`retrieve(query, k)` by BM25 over that record), `Memory` (agent-authored notebook, keyed by namespace). The real argument in #4773, that load/save "only moves the full-history replay behind a protocol" and the right shape is `retrieve(query, k)`, is correct about load/save alone and wrong in its conclusion: retrieval needs a corpus, and the log is the corpus. The cleanest statement of the boundary in either thread is @IgnazioDS's: `message_history` for verbatim replay, a retrieval layer for durable facts, and conflating the two is why the interface has been hard to get right. Recommendation is to reframe #4773 onto the retrieval layer rather than ship it as specified, since its method set and its identifier both collide with this.

**6. Async only.** LangGraph is the cautionary tale and it got worse after launch (~12 sync methods plus a full async mirror). OpenAI's `Session` is async-only and has eight backends, so async-only demonstrably does not suppress an ecosystem. Harness `StepStore` is already async-only, and its file and SQLite backends implement sync bodies internally behind async methods, which is the pattern that serves sync-first users without doubling the protocol. `run_sync` users are unaffected.

**7. Provider-managed state stays a model setting.** It already ships as `openai_conversation_id`. It cannot honestly implement the protocol: `get_messages` has no implementation, because the provider gives you continuation and not readable history. It is not portable, it is unavailable under Zero Data Retention, and it does not even reduce billed input tokens. OpenAI can model theirs as a backend because their `Session` only ever feeds their own API; ours is provider-agnostic, so the same move would be a category error. What is worth doing is documenting the composition, since a store and `openai_conversation_id='auto'` coexist fine with the store as source of truth.

</details>

<details>
<summary><b>Staged plan</b></summary>

**Stage 0, docs only.** #7821. Satisfies the findability half at zero design risk, correct however the forks land, and deliberately not blocking on them.

**Stage 1, core seam.** `SessionStore` protocol and ABC, `InMemorySessionStore`, a `SessionPersistence` capability, and the doctrine rewrite from fork 3. Blocked on forks 1, 2, 6 and maintainer sign-off.

**Stage 2, Harness adapter.** `StepStore` to `SessionStore`, so the shipped SQLite, file and MongoDB backends serve the simple case with no new backend code.

**Stage 3, branching.** Revisit via pydantic-ai-harness#321 once a real consumer exists. Nothing in Stage 1 forecloses it provided fork 2 lands on reserving the pointer.

**Never, under this proposal:** graph-state checkpointing (#149). That is dmontagu's objection and it is a good one.

</details>

**Comment by RemanenetSpy** [comment]
As @ferhimedamine pointed out, the core challenge with external message persistence isn't serializing `ModelMessage` objects—it's what happens when history reaches thousands of turns across sessions and dumping full chat logs overflows the context window.

Storing raw chat messages in JSONB creates an unindexable token swamp. The standard architectural solution is a **two-tier memory model**:

1. **Turn Calendar (`turns` table):** Stores raw message payloads for complete sequential replay when needed.
2. **SVO Event Ledger (`events` table):** Asynchronously extracts discrete Subject-Verb-Object assertions with `valid_from` / `valid_to` bounds and pgvector embeddings.

```python
# On session start, inject only the compact active fact ledger (< 2k chars)
active_facts = await db.fetch_all("""
    SELECT subject, verb, object FROM events 
    WHERE source_id = :user_id AND valid_to IS NULL
    ORDER BY timestamp DESC LIMIT 20
""")
# Agent starts with exact current beliefs without loading thousands of raw chat turns
```

This lets `pydantic-ai` support unbounded cross-session persistence while keeping startup token overhead strictly constant (O(1)).

**Comment by dsfaccini** [comment]
> This comment was posted by Claude Code using claude-opus-5 on behalf of David.

That's a separate concern from this issue. #530 is about **persisting and reloading `ModelMessage` history** across processes — a store protocol plus adapters. Deciding *what* of that history reaches the context window is a layer above it, and one we already ship: [`pydantic-ai-harness`](https://github.com/pydantic/pydantic-ai-harness) has a `memory` capability (with a Postgres backend) and summarizing / tiered / sliding-window compaction, all of which compose with whatever store holds the messages.

Coupling the core session abstraction to one memory architecture is what we're specifically trying not to do. Happy to discuss a fact-ledger memory backend on the harness tracker.

**Comment by RemanenetSpy** [comment]
@dsfaccini Understood and agreed—clean separation of concerns. Decoupling the low-level `ModelMessage` store protocol in core from higher-level memory projections in `pydantic-ai-harness` makes total sense. We will bring the SVO Fact Ledger backend proposal over to the harness tracker.

**Comment by 1picassoai** [comment]
Field note from building exactly this class of thing: persisting the MESSAGES turns out to be the less useful half. Raw message history grows without bound, re-costs tokens on every rehydration, and most of it is repetition - I measured my own agent sessions and ~70% of processed context was re-sent history (anyone can check theirs: `npx github:1picassoai/juvina-token-bill` reads local logs, nothing leaves the machine).

What worked better than a Redis message cache: persist MEANING, not transcript - encode decisions/facts as they happen into a local store, and on resume recall a compact constructed answer (a few hundred tokens) instead of reloading history. State survives restarts and store swaps because it never lived in the message list to begin with.

For pydantic-ai specifically the AbstractMemoryStore RFC (#4773) feels like the right seam - a message-store backend AND a memory-engine backend could both sit behind it. Disclosure: I build juvina.ai (local memory engine over MCP); sharing the pattern, not prescribing the backend.

**Comment by DouweM** [comment]
Thanks @dsfaccini — the survey is what made this decidable. Ruling on fork 1, plus what Samuel's input on today's open source sync settles for two of the others.

## Fork 1: a log, and an envelope beside it

The fork is posed as log *or* resumable run state. Having looked at what people actually ship, I don't think that binary survives contact — nobody picks a side:

| | Payload | Stored beside it |
|---|---|---|
| [OpenAI Agents SDK](https://openai.github.io/openai-agents-python/sessions/) | flat item log | usage, via an explicit `store_run_usage(result)` |
| [LangGraph](https://docs.langchain.com/oss/python/langgraph/persistence) | full state snapshot per step | parent pointer |
| [Code Puppy](https://github.com/mpfaffenberger/code_puppy) | message envelope (`ModelMessagesTypeAdapter`) | `_meta` sidecar: `total_tokens`, `message_count`, `scope_key` |
| [csp-bot](https://github.com/Point72/csp-bot) (Point72) | message list | `last_active`, `bot_response_id` |
| [pydantic-deepagents](https://github.com/vstorm-co/pydantic-deepagents) | messages | `turn`, `label`, `metadata` |
| Harness `StepPersistence` | event log | snapshots, tool-effect ledger |

Five of six store a message log. Five of six store something next to it. So: **the payload is an append-only log of `ModelMessage`, and an envelope carries what the log provably cannot.** That keeps everything fork 1 was defending — no graph-state checkpointing, @dmontagu's objection still cleared, append-only, branchable later. It only rejects the idea that the envelope is the *alternative* to the log rather than its companion.

What goes in the envelope, and why each item earns its place:

- **`usage`.** Not one of the six derives it from the log. Code Puppy carries a 415-line `token_usage.py` that re-*estimates* tokens from raw text — messages, tool JSON schemas, MCP server schemas, system-prompt overhead — to rebuild a context-window gauge, because no real `RunUsage` survives into a stored session. On our side it is *almost* derivable: every `RequestUsage` field and `requests` reconstruct exactly from `ModelResponse.usage`, but `tool_calls` does not (it increments on executed local tools, skipping retried ones and provider-side native calls). And deriving has a worse failure than being approximate: it shrinks when you trim history, so a long conversation's cost cap silently resets itself.
- **`conversation_id`**, which we already resolve, inherit and fork.
- **A format version.** Both third-party implementations defend against us — Code Puppy stamps the installed pydantic-ai version (read without importing the package), csp-bot hard-fails on its own `SCHEMA_VERSION` mismatch. We do guarantee that data serialized by an older version still deserializes, and we say so in this tracker, but nothing in the API expresses it, so people write their own guard.
- **The deferred-resume state, only while a run is paused.** `DeferredToolResults` come from the caller, not the run, so the log cannot hold them by construction — and nested sub-agent history (#4302) has nowhere to live at all today.

Your day-one ask for an entry-kind discriminator is what lets that paused-run entry join the log later without a migration, so it should ship regardless of when we use it.

Two more things every one of these got that no proposal in this thread has: **a retention policy** (count cap, TTL, `remove_oldest()` in the store protocol itself) and **a scope key that isn't `conversation_id`** (a resolved cwd, `command:user:channel`, a namespace). `conversation_id` names the dialogue; it doesn't say whose or where.

## What today's sync settles

Samuel asked for "a go-to memory implementation, both memory and just storing messages so you can retrieve them easily", given to an agent "with one capability".

- **Fork 5** — sessions vs memory, "both, layered" — holds internally, but not on the surface. One capability, not two the user assembles.
- **The in-memory default in Stage 1 isn't enough.** Samuel wants Turso locally "so it just works", Postgres if we host one, and explicitly "I don't think it should just be file system". A default that doesn't persist on first use fails the ask. Worth noting that all three agent-memory implementations I looked at — Harness `Memory`, pydantic-deepagents, and the file-per-session stores in this thread — are files.

Turso is a good fit for that default, and it doesn't disturb your core/Harness split. [`turso.connect("app.db")`](https://docs.turso.tech/sdk/python/quickstart) is a plain local file with no account and no network — it is a SQLite fork, so the zero-config default behaves like a real database rather than a directory of JSON. The same library then upgrades to an embedded replica (`turso.sync.connect(..., remote_url=..., auth_token=...)` with explicit `push()`/`pull()`), so local to hosted is a connection change rather than a rewrite of the storage layer. The package is compiled with wheels across supported Pythons and no Python dependencies, which makes it a reasonable Harness dependency and a poor one for `pydantic-ai-slim`.

So Stage 2 stays as you proposed: the protocol in core, the shipped backends in Harness. Harness is the package essentially everyone pulls in alongside Pydantic AI, so "one capability" and "backends live in Harness" are the same sentence, not competing ones.

Forks 2, 3, 6 and 7 I'm taking as recommended.

<sub>Posted by Claude Code (claude-opus-5) on behalf of Douwe.</sub>

**Comment by kevin-lozada-santos** [comment]
@izzyacademy Persisted message history is a good test surface because the same conversation has to survive serialization, external storage, reload, and the next model run without losing ordering or tool-call identity. If you have a public or sanitized fixture, would you try a free, read-only Brain Scanner run yourself at https://brainscanner.dev/? Come back 12–24 hours later for a second check and tell me whether the map keeps those persistence edges connected. No credentials, private data, or repo changes.
